### ISSUE TITLE
Avoid spurious recompilations when unrelated constructor changes

### ISSUE THREAD

**Issue by smarter** [sbt/zinc#97]
`A.scala`:

``` scala
class A(a: Int) // change Int to String, B.scala should not be recompiled
object A { val x = 3 }
```

`B.scala`:

``` scala
class B { val y = A.x }
```

`B` does not use the constructor of `A`, but every constructor has the same name `<init>`, so if we change a constructor in a class, then all dependencies of that class are inevitably recompiled. This should be easy to fix: just special case constructors and prepend the class name to the method name, like `A<init>`, `B<init>`.

(Note that this change will break value-class-underlying if #95 gets merged (see the PR description), this can be fixed by including the underlying val of a value class in its name hash)

**Comment by gkossakowski** [comment]
The same problem exists for apply, copy, toString, etc. Why special case constructors?

**Comment by smarter** [comment]
For other methods you have the added complication of implicits and imports, even if an unrelated `apply` method changes, you might need to recompile `foo.apply` because it relies on implicit resolution. That can never happen for `<init>`.

**Comment by gkossakowski** [comment]
Fair enough. I'm ok with a change like that. I'd maybe use `ClassName::<init>` syntax to make it a bit more clear that the name has been constructed intentionally. That's a minor point, though.

**Comment by gkossakowski** [comment]
PS. I merged #95.
